File: glim_ws/glim_converters.py
#!/usr/bin/env python3
"""
GLIM converter nodes:
  1. Livox CustomMsg -> PointCloud2 (with per-point time offsets)
  2. sensor_msgs/NavSatFix -> geometry_msgs/PoseWithCovarianceStamped (ENU local coords)
  3. /glim_ros/map QoS bridge (TRANSIENT_LOCAL -> VOLATILE) for octomap_server

Usage:
  python3 glim_converters.py
  python3 glim_converters.py --ros-args \
    -p lidar_input_topic:=/livox/lidar \
    -p lidar_output_topic:=/livox/lidar_pointcloud2 \
    -p gps_input_topic:=/gps/fix \
    -p gps_output_topic:=/gps/fix_pose \
    -p map_input_topic:=/glim_ros/map \
    -p map_output_topic:=/glim_ros/map_volatile
"""

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy, HistoryPolicy
import math
import numpy as np

from sensor_msgs.msg import NavSatFix, PointCloud2, PointField
from geometry_msgs.msg import PoseWithCovarianceStamped
from livox_ros_driver2.msg import CustomMsg


# WGS84 ellipsoid parameters
WGS84_A = 6378137.0
WGS84_E2 = 0.00669437999014


def geodetic_to_ecef(lat, lon, alt):
    lat_r = math.radians(lat)
    lon_r = math.radians(lon)
    N = WGS84_A / math.sqrt(1 - WGS84_E2 * math.sin(lat_r) ** 2)
    x = (N + alt) * math.cos(lat_r) * math.cos(lon_r)
    y = (N + alt) * math.cos(lat_r) * math.sin(lon_r)
    z = (N * (1 - WGS84_E2) + alt) * math.sin(lat_r)
    return x, y, z


def ecef_to_enu(ecef, origin_ecef, origin_lat, origin_lon):
    lat_r = math.radians(origin_lat)
    lon_r = math.radians(origin_lon)
    dx = ecef[0] - origin_ecef[0]
    dy = ecef[1] - origin_ecef[1]
    dz = ecef[2] - origin_ecef[2]
    e = -math.sin(lon_r) * dx + math.cos(lon_r) * dy
    n = (-math.sin(lat_r) * math.cos(lon_r) * dx
         - math.sin(lat_r) * math.sin(lon_r) * dy
         + math.cos(lat_r) * dz)
    u = (math.cos(lat_r) * math.cos(lon_r) * dx
         + math.cos(lat_r) * math.sin(lon_r) * dy
         + math.sin(lat_r) * dz)
    return e, n, u


class GlimConverters(Node):
    def __init__(self):
        super().__init__('glim_converters')

        # --- Parameters ---
        self.declare_parameter('lidar_input_topic',  '/livox/lidar')
        self.declare_parameter('lidar_output_topic', '/livox/lidar_pointcloud2')
        self.declare_parameter('gps_input_topic',    '/gps/fix')
        self.declare_parameter('gps_output_topic',   '/gps/fix_pose')
        self.declare_parameter('map_input_topic',    '/glim_ros/map')
        self.declare_parameter('map_output_topic',   '/glim_ros/map_volatile')

        lidar_in  = self.get_parameter('lidar_input_topic').value
        lidar_out = self.get_parameter('lidar_output_topic').value
        gps_in    = self.get_parameter('gps_input_topic').value
        gps_out   = self.get_parameter('gps_output_topic').value
        map_in    = self.get_parameter('map_input_topic').value
        map_out   = self.get_parameter('map_output_topic').value

        # --- LiDAR converter ---
        self.lidar_sub = self.create_subscription(CustomMsg, lidar_in, self.lidar_callback, 10)
        self.lidar_pub = self.create_publisher(PointCloud2, lidar_out, 10)

        # Preallocate fields list (same every time)
        self._fields = [
            PointField(name='x',         offset=0,  datatype=PointField.FLOAT32, count=1),
            PointField(name='y',         offset=4,  datatype=PointField.FLOAT32, count=1),
            PointField(name='z',         offset=8,  datatype=PointField.FLOAT32, count=1),
            PointField(name='intensity', offset=12, datatype=PointField.FLOAT32, count=1),
            PointField(name='time',      offset=16, datatype=PointField.FLOAT32, count=1),
        ]

        # --- GPS converter ---
        self.gps_sub = self.create_subscription(NavSatFix, gps_in, self.gps_callback, 10)
        self.gps_pub = self.create_publisher(PoseWithCovarianceStamped, gps_out, 10)
        self.gps_origin_ecef = None
        self.gps_origin_lat  = None
        self.gps_origin_lon  = None

        # --- Map QoS bridge (TRANSIENT_LOCAL -> VOLATILE for octomap_server) ---
        transient_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.TRANSIENT_LOCAL,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )
        volatile_qos = QoSProfile(
            reliability=ReliabilityPolicy.BEST_EFFORT,
            durability=DurabilityPolicy.VOLATILE,
            history=HistoryPolicy.KEEP_LAST,
            depth=10
        )
        self.map_sub = self.create_subscription(
            PointCloud2, map_in, self.map_callback, transient_qos)
        self.map_pub = self.create_publisher(PointCloud2, map_out, volatile_qos)

        self.get_logger().info(f'[LiDAR] {lidar_in} -> {lidar_out}')
        self.get_logger().info(f'[GPS]   {gps_in} -> {gps_out}')
        self.get_logger().info(f'[MAP]   {map_in} -> {map_out} (QoS bridge)')

    # -------------------------------------------------------------------------
    # LiDAR: CustomMsg -> PointCloud2 (numpy vectorized)
    # -------------------------------------------------------------------------
    def lidar_callback(self, msg: CustomMsg):
        n = msg.point_num
        if n == 0:
            return

        points = msg.points

        # Vectorized extraction using numpy
        arr = np.empty((n, 5), dtype=np.float32)
        arr[:, 0] = [p.x            for p in points]
        arr[:, 1] = [p.y            for p in points]
        arr[:, 2] = [p.z            for p in points]
        arr[:, 3] = [p.reflectivity for p in points]
        arr[:, 4] = [p.offset_time * 1e-9 for p in points]  # ns -> s

        pc2 = PointCloud2()
        pc2.header      = msg.header
        pc2.height      = 1
        pc2.width       = n
        pc2.fields      = self._fields
        pc2.is_bigendian = False
        pc2.point_step  = 20  # 5 * float32
        pc2.row_step    = 20 * n
        pc2.data        = arr.tobytes()
        pc2.is_dense    = False
        self.lidar_pub.publish(pc2)

    # -------------------------------------------------------------------------
    # GPS: NavSatFix -> PoseWithCovarianceStamped (ENU)
    # -------------------------------------------------------------------------
    def gps_callback(self, msg: NavSatFix):
        if msg.status.status < 0:
            return

        lat, lon, alt = msg.latitude, msg.longitude, msg.altitude
        ecef = geodetic_to_ecef(lat, lon, alt)

        if self.gps_origin_ecef is None:
            self.gps_origin_ecef = ecef
            self.gps_origin_lat  = lat
            self.gps_origin_lon  = lon
            self.get_logger().info(
                f'[GPS] origin set: lat={lat:.6f} lon={lon:.6f} alt={alt:.2f}')

        e, n, u = ecef_to_enu(ecef, self.gps_origin_ecef,
                               self.gps_origin_lat, self.gps_origin_lon)

        pose_msg = PoseWithCovarianceStamped()
        pose_msg.header = msg.header
        pose_msg.header.frame_id = 'map'
        pose_msg.pose.pose.position.x = e
        pose_msg.pose.pose.position.y = n
        pose_msg.pose.pose.position.z = u
        pose_msg.pose.pose.orientation.w = 1.0

        cov = [0.0] * 36
        if msg.position_covariance_type > 0:
            cov[0]  = msg.position_covariance[0]
            cov[1]  = msg.position_covariance[1]
            cov[2]  = msg.position_covariance[2]
            cov[6]  = msg.position_covariance[3]
            cov[7]  = msg.position_covariance[4]
            cov[8]  = msg.position_covariance[5]
            cov[12] = msg.position_covariance[6]
            cov[13] = msg.position_covariance[7]
            cov[14] = msg.position_covariance[8]
        else:
            cov[0] = cov[7] = cov[14] = 1.0

        pose_msg.pose.covariance = cov
        self.gps_pub.publish(pose_msg)

    # -------------------------------------------------------------------------
    # Map QoS bridge: TRANSIENT_LOCAL -> VOLATILE
    # -------------------------------------------------------------------------
    def map_callback(self, msg: PointCloud2):
        self.map_pub.publish(msg)


def main():
    rclpy.init()
    node = GlimConverters()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: glim_ws/src/glim_converters/src/glim_converters.cpp
#include <cmath>
#include <memory>
#include <vector>
#include <array>

#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/point_cloud2.hpp>
#include <sensor_msgs/msg/point_field.hpp>
#include <sensor_msgs/msg/nav_sat_fix.hpp>
#include <geometry_msgs/msg/pose_with_covariance_stamped.hpp>
#include <livox_ros_driver2/msg/custom_msg.hpp>

static constexpr double WGS84_A  = 6378137.0;
static constexpr double WGS84_E2 = 0.00669437999014;

static std::array<double, 3> geodetic_to_ecef(double lat, double lon, double alt)
{
  const double lat_r = lat * M_PI / 180.0;
  const double lon_r = lon * M_PI / 180.0;
  const double N = WGS84_A / std::sqrt(1.0 - WGS84_E2 * std::sin(lat_r) * std::sin(lat_r));
  return {
    (N + alt) * std::cos(lat_r) * std::cos(lon_r),
    (N + alt) * std::cos(lat_r) * std::sin(lon_r),
    (N * (1.0 - WGS84_E2) + alt) * std::sin(lat_r)
  };
}

static std::array<double, 3> ecef_to_enu(
  const std::array<double, 3>& ecef,
  const std::array<double, 3>& origin_ecef,
  double origin_lat, double origin_lon)
{
  const double lat_r = origin_lat * M_PI / 180.0;
  const double lon_r = origin_lon * M_PI / 180.0;
  const double dx = ecef[0] - origin_ecef[0];
  const double dy = ecef[1] - origin_ecef[1];
  const double dz = ecef[2] - origin_ecef[2];
  return {
    -std::sin(lon_r) * dx + std::cos(lon_r) * dy,
    -std::sin(lat_r) * std::cos(lon_r) * dx
      - std::sin(lat_r) * std::sin(lon_r) * dy
      + std::cos(lat_r) * dz,
    std::cos(lat_r) * std::cos(lon_r) * dx
      + std::cos(lat_r) * std::sin(lon_r) * dy
      + std::sin(lat_r) * dz
  };
}

class GlimConverters : public rclcpp::Node
{
public:
  GlimConverters() : Node("glim_converters"), gps_origin_set_(false)
  {
    declare_parameter("lidar_input_topic",  "/livox/lidar");
    declare_parameter("lidar_output_topic", "/livox/lidar_pointcloud2");
    declare_parameter("gps_input_topic",    "/gps/fix");
    declare_parameter("gps_output_topic",   "/gps/fix_pose");
    declare_parameter("gps_enu_yaw_deg",    214.8);
    declare_parameter("gps_offset_x",       0.0);
    declare_parameter("gps_offset_y",       0.1);
    declare_parameter("gps_offset_z",       0.0);

    const auto lidar_in  = get_parameter("lidar_input_topic").as_string();
    const auto lidar_out = get_parameter("lidar_output_topic").as_string();
    const auto gps_in    = get_parameter("gps_input_topic").as_string();
    const auto gps_out   = get_parameter("gps_output_topic").as_string();

    lidar_sub_ = create_subscription<livox_ros_driver2::msg::CustomMsg>(
      lidar_in, 10,
      std::bind(&GlimConverters::lidar_callback, this, std::placeholders::_1));
    lidar_pub_ = create_publisher<sensor_msgs::msg::PointCloud2>(lidar_out, 10);

    fields_.resize(5);
    fields_[0].name = "x";         fields_[0].offset = 0;  fields_[0].datatype = sensor_msgs::msg::PointField::FLOAT32; fields_[0].count = 1;
    fields_[1].name = "y";         fields_[1].offset = 4;  fields_[1].datatype = sensor_msgs::msg::PointField::FLOAT32; fields_[1].count = 1;
    fields_[2].name = "z";         fields_[2].offset = 8;  fields_[2].datatype = sensor_msgs::msg::PointField::FLOAT32; fields_[2].count = 1;
    fields_[3].name = "intensity"; fields_[3].offset = 12; fields_[3].datatype = sensor_msgs::msg::PointField::FLOAT32; fields_[3].count = 1;
    fields_[4].name = "time";      fields_[4].offset = 16; fields_[4].datatype = sensor_msgs::msg::PointField::FLOAT32; fields_[4].count = 1;

    gps_sub_ = create_subscription<sensor_msgs::msg::NavSatFix>(
      gps_in, 10,
      std::bind(&GlimConverters::gps_callback, this, std::placeholders::_1));
    gps_pub_ = create_publisher<geometry_msgs::msg::PoseWithCovarianceStamped>(gps_out, 10);

    RCLCPP_INFO(get_logger(), "[LiDAR] %s -> %s", lidar_in.c_str(), lidar_out.c_str());
    RCLCPP_INFO(get_logger(), "[GPS]   %s -> %s", gps_in.c_str(), gps_out.c_str());
  }

private:
  void lidar_callback(const livox_ros_driver2::msg::CustomMsg::SharedPtr msg)
  {
    const uint32_t n = msg->point_num;
    if (n == 0) return;

    constexpr uint32_t POINT_STEP = 20;
    std::vector<uint8_t> data(n * POINT_STEP);

    float* ptr = reinterpret_cast<float*>(data.data());
    for (uint32_t i = 0; i < n; ++i) {
      const auto& p = msg->points[i];
      ptr[i * 5 + 0] = p.x;
      ptr[i * 5 + 1] = p.y;
      ptr[i * 5 + 2] = p.z;
      ptr[i * 5 + 3] = static_cast<float>(p.reflectivity);
      ptr[i * 5 + 4] = static_cast<float>(p.offset_time) * 1e-9f;
    }

    sensor_msgs::msg::PointCloud2 out;
    out.header       = msg->header;
    out.height       = 1;
    out.width        = n;
    out.fields       = fields_;
    out.is_bigendian = false;
    out.point_step   = POINT_STEP;
    out.row_step     = POINT_STEP * n;
    out.data         = std::move(data);
    out.is_dense     = false;

    lidar_pub_->publish(out);
  }

  void gps_callback(const sensor_msgs::msg::NavSatFix::SharedPtr msg)
  {
    if (msg->status.status < 0) return;

    const double lat = msg->latitude;
    const double lon = msg->longitude;
    const double alt = msg->altitude;
    const auto ecef = geodetic_to_ecef(lat, lon, alt);

    if (!gps_origin_set_) {
      gps_origin_ecef_ = ecef;
      gps_origin_lat_  = lat;
      gps_origin_lon_  = lon;
      gps_origin_set_  = true;
      RCLCPP_INFO(get_logger(), "[GPS] origin set: lat=%.6f lon=%.6f alt=%.2f", lat, lon, alt);
    }

    const auto enu = ecef_to_enu(ecef, gps_origin_ecef_, gps_origin_lat_, gps_origin_lon_);

    const double yaw   = get_parameter("gps_enu_yaw_deg").as_double() * M_PI / 180.0;
    const double cos_y = std::cos(yaw);
    const double sin_y = std::sin(yaw);

    const double ox = get_parameter("gps_offset_x").as_double();
    const double oy = get_parameter("gps_offset_y").as_double();
    const double oz = get_parameter("gps_offset_z").as_double();

    geometry_msgs::msg::PoseWithCovarianceStamped pose_msg;
    pose_msg.header          = msg->header;
    pose_msg.header.frame_id = "map";
    pose_msg.pose.pose.position.x    = cos_y * enu[0] + sin_y * enu[1] - ox;
    pose_msg.pose.pose.position.y    = -sin_y * enu[0] + cos_y * enu[1] - oy;
    pose_msg.pose.pose.position.z    = enu[2] - oz;

    const double yaw_corr = get_parameter("gps_enu_yaw_deg").as_double() * M_PI / 180.0;
    pose_msg.pose.pose.orientation.z = std::cos(yaw_corr / 2.0);
    pose_msg.pose.pose.orientation.w = std::sin(yaw_corr / 2.0);

    if (msg->position_covariance_type > 0) {
      pose_msg.pose.covariance[0]  = msg->position_covariance[0];
      pose_msg.pose.covariance[1]  = msg->position_covariance[1];
      pose_msg.pose.covariance[2]  = msg->position_covariance[2];
      pose_msg.pose.covariance[6]  = msg->position_covariance[3];
      pose_msg.pose.covariance[7]  = msg->position_covariance[4];
      pose_msg.pose.covariance[8]  = msg->position_covariance[5];
      pose_msg.pose.covariance[12] = msg->position_covariance[6];
      pose_msg.pose.covariance[13] = msg->position_covariance[7];
      pose_msg.pose.covariance[14] = msg->position_covariance[8];
    } else {
      pose_msg.pose.covariance[0] = pose_msg.pose.covariance[7] = pose_msg.pose.covariance[14] = 1.0;
    }

    gps_pub_->publish(pose_msg);
  }

  rclcpp::Subscription<livox_ros_driver2::msg::CustomMsg>::SharedPtr lidar_sub_;
  rclcpp::Publisher<sensor_msgs::msg::PointCloud2>::SharedPtr         lidar_pub_;
  std::vector<sensor_msgs::msg::PointField> fields_;

  rclcpp::Subscription<sensor_msgs::msg::NavSatFix>::SharedPtr                gps_sub_;
  rclcpp::Publisher<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr  gps_pub_;
  bool gps_origin_set_;
  std::array<double, 3> gps_origin_ecef_{};
  double gps_origin_lat_{0.0}, gps_origin_lon_{0.0};
};

int main(int argc, char** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<GlimConverters>());
  rclcpp::shutdown();
  return 0;
}